### ISSUE TITLE
"Sonic 2 & Knuckles" black screen

### ISSUE THREAD

**Issue by xenyou** [MiSTer-devel/Genesis_MiSTer#220]
maybe it's the same issue as #145.

Sonic2 + "Sonic and Knuckles" does not boot on Mister. It works fine in OpenEmu.

And I found some report that it worked on Mister before.
https://www.atari-forum.com/viewtopic.php?p=374641#p374641

I have the same ROM image in the above forum.
Is there some regressions?

Information
BIN CRC: 2AC1E7C6
bin file md5: 3e5e4b18d035775b916a06f2b3dc5031
region: JP

**Comment by sorgelig** [comment]
I have this one:
MD5: bdd13934f4f31bf4d849ef8597d4a07c 
"Sonic & Knuckles + Sonic The Hedgehog 2 (World) (Rev A) EverDrive Fixed v1.1 SmokeMonster.bin"

It works. Not sure what "fixed" means. I will look for other versions.

**Comment by birdybro** [comment]
There is a bios could be used eventually one day to resolve this, but if there's already a hack that works it's kind of a moot point.

https://docs.libretro.com/library/genesis_plus_gx/#bios

These lock-on roms seem to be important, I think sk2chip.bin (b4e76e416b887f4e7413ba76fa735f16) is the one that would fix it, and it's possible the hack implements that chip or uses a different mapper. These lock-on roms I think are required for genesis-plus-gx to work with it, or used to be.

**Comment by Missingno255** [comment]
> **Important Edit**: https://retrocomputing.stackexchange.com/a/1529 - This is the most thorough explanation of how it works. Sonic and Knuckles has a sonic2 patched version of the game on the cartridge itself specifically, and defers to that game. This is what sk2chip.bin is. I'm assuming the hacked version by smokemonster is just moving the mapping around so whatever emulator (in that case the Everdrive) doesn't have to recreate the hardware to do this mapping and carry support for sk2chip.bin loading conditionally.
> 
> Old comment: There is a bios could be used eventually one day to resolve this, but if there's already a hack that works it's kind of a moot point.
> 
> https://docs.libretro.com/library/genesis_plus_gx/#bios
> 
> These lock-on roms seem to be important, I think sk2chip.bin (b4e76e416b887f4e7413ba76fa735f16) is the one that would fix it, and it's possible the hack implements that chip or uses a different mapper. These lock-on roms I think are required for genesis-plus-gx to work with it, or used to be.
> 
> EDIT: https://github.com/ekeeke/Genesis-Plus-GX/blob/master/core/cart_hw/md_cart.c#L730-L772
> 
> https://github.com/ekeeke/Genesis-Plus-GX/blob/master/core/cart_hw/md_cart.c#L842-L857
> 
> https://github.com/ekeeke/Genesis-Plus-GX/blob/master/core/cart_hw/md_cart.c#L1009-L1019
> 
> https://segaretro.org/Lock-On_Technology#Lock-On_with_Sonic_the_Hedgehog_2
> 
> > When Sonic 2 is connected, the cart combines the Sonic & Knuckles ROM, the Sonic 2 ROM, and a 256K ROM chip to boot into a new game entitled [Knuckles in Sonic 2](https://segaretro.org/index.php?title=Knuckles_in_Sonic_2&action=edit&redlink=1). This chip contains 'patch code' such as object placement data (for example, adding a 1-Up to the top of a building in the Chemical Plant Zone), Knuckles sprites and mappings, all code needed for the game to function, and some [Casino Night Zone](https://info.sonicretro.org/Casino_Night_Zone) collision data. This chip alone is available in 'dumped' form; however, it is unplayable in emulators, as it is not a complete game in and of itself, being more analogous to an IPS patch.
> 
> > Multiple prototypes of this patch chip were released in the [February 23, 2008 Proto Release](https://info.sonicretro.org/February_23,_2008_Proto_Release).
> 
> https://info.sonicretro.org/Knuckles_the_Echidna_in_Sonic_the_Hedgehog_2#Technical_information
> 
> > When Sonic & Knuckles detects that Sonic 2 is locked on (it checks for Revisions 00, 01 and 02), it transfers control to a special hidden 256 KB ROM chip, often referred to as the UPMEM. This contains the code for the entire game, as well as some data (e.g. the graphics of Knuckles in the Special Stage). The only code run from the locked-on Sonic 2 ROM is the sound driver, although much of the game's data is accessed from the Sonic 2 ROM instead of being duplicated in the UPMEM.
> 
> Even a youtube video on it! --> https://www.youtube.com/watch?v=9-4SApbrYH0 :)

It doesn't need a mapper. The No-Intro Romset ROM image for Knuckles in Sonic 2 needs to have both Sonic 2 and Sonic & Knuckles, along with the 256KB UPMEM chip, sk2chip.bin, pre-patched, as it won't work otherwise, obviously. That's the reason for the black screen.

**Comment by birdybro** [comment]
>It doesn't need a mapper. The No-Intro Romset ROM image for Knuckles in Sonic 2 needs to have both Sonic 2 and Sonic & Knuckles, along with the 256KB UPMEM chip, sk2chip.bin, pre-patched, as it won't work otherwise, obviously. That's the reason for the black screen.

I didn't say it needs a mapper. Just saying what's possible. OP's hash matches the dump that excludes the lock-on chip, yes. The Everdrive fix version includes that 256KB from the UPMEM.

**Comment by Missingno255** [comment]
> > It doesn't need a mapper. The No-Intro Romset ROM image for Knuckles in Sonic 2 needs to have both Sonic 2 and Sonic & Knuckles, along with the 256KB UPMEM chip, sk2chip.bin, pre-patched, as it won't work otherwise, obviously. That's the reason for the black screen.
> 
> I didn't say it needs a mapper. Just saying what's possible. OP's hash matches the dump that excludes the lock-on chip, yes. The Everdrive fix version includes that 256KB from the UPMEM. I'm saying it would resolve the issue if loading of sk2chip.bin were implemented like other emulators. It's obviously not worth the effort overall, but just documenting it somewhat.

Yeah, it would be nice to have that built in. Honestly though, No-Intro needs to pre-patch everything if they want proper preservation of KiS2 and for it to work without having to make emulator authors to add extra emulation of a chip that should have been patched in the first place. Real hardware compatibility on flash carts would happen right off that bat too, as a bonus. :)

**Comment by sorgelig** [comment]
This is very specific just to single game. And there is a fixed dump which doesn't exclude any content.

**Comment by tjanas** [comment]
The "EverDrive Fixed v1.1 SmokeMonster.bin" (md5: bdd13934f4f31bf4d849ef8597d4a07c) alters four bytes in the concatenated S&K + Sonic2(RevA) + lock-on-chip file (md5: 3e5e4b18d035775b916a06f2b3dc5031):
```
Address 0000018E: 0xDFB3 -> 0x8603
Address 000001A5: 0x1F -> 0x33
Address 000003B5: 0x01 -> 0x02
```
However, I cannot explain the reasoning.
![image](https://user-images.githubusercontent.com/20304619/189401369-83ac3fbe-2a57-4154-88c7-8959f35f0739.png)
![image](https://user-images.githubusercontent.com/20304619/189401602-9b3e62a2-566b-4d02-94e7-b51cab9c47db.png)

**Comment by xenyou** [comment]
Thanks for many information.
I've just edited my dump image with binary editor, and it works fine!!

So Mister currently needs this modification to play this game. right?
